Details to motivation (#3)
tc39-transfer/proposal-object-pick-or-omit: README.md

## `Object.{pick, omit}`
> ECMAScript Proposal, specs, and reference implementation for `Object.pick`, `Object.omit`.

Spec drafted by [@Aleen](https://github.com/aleen42).

### Motivation

Let us consider few scenarios from the real would to understand what we are trying to solve in this propsoal. 

* On `MouseEvent` we are intreseted on `'ctrlKey', 'shiftKey', 'altKey', 'metaKey'` events only.
* We have a `configObject` and we need `['dependencies', 'devDependencies', 'peerDependencies']` from it. 
* We have an `optionsBag`and we would allow on `['shell', 'env', 'extendEnv', 'uid', 'gid']` on it.
* From a `req.body` we want to extract `['name', 'company', 'email', 'password']`
* Checking if a component `shouldReload` by extracting `compareKeys` from `props` and compare it with `prevProps`. 
* Say we have a `depsObject` and we need to ignore all `@internal/packages` from it.
* We have `props` from which we need to remove `[‘_csrf’, ‘_method’]`
* We need to construct a `newModelData` by removing `action.deleted` from `({ ...state.models, ...action.update })`

Well, you see life is all about `pick`ing what we want and `omit`ing what we don't! 

> Would life be easier if the language provided a convinent method to help us during similar scenarios?

Reviewer: ```suggestion
Would life be easier if the language provided a convenient method to help us during similar scenarios?
```

Now, one might argue saying we can implement `pick` and `omit` as below:

```js

> const pick = keys => Object.fromEntries(

Reviewer: ```suggestion
const pick = (obj, keys) => Object.fromEntries(
```

    keys.map(k => obj.hasOwnProperty(k) && [k, obj[k]]).filter(x => x)
);

/*
We can also use a Destructuring assignment
const { authKey, ...toLog } = userInfo;
*/
```

```js

> const omit = keys => Object.fromEntries(

Reviewer: ```suggestion
const omit = (obj, keys) => Object.fromEntries(
```

    keys.map(k => !obj.hasOwnProperty(k) && [k, obj[k]]).filter(x => x)
);
```

The major challenges we see with the above implementations:

* It is not ergonomic 

> * We can't omit for unknown values

Reviewer: this applies to destructuring, but not to your custom function, which take an array

Author: Yes, destructuring is mentioned in the comments section in the above implementation

> We can read more about such use-cases and challenges in [es.discourse](https://es.discourse.group/t/object-prototype-pick-object-prototype-omit/515).

Reviewer: there's like a dozen threads in the discourse and es-discuss; i'd link to them all

With that in mind would it not be easier if we had `Object.pick` and `Object.omit` static methods?!

> Let us now discuss about what the sytanx of such useful method would be?

Reviewer: ```suggestion
Let us now discuss about what the API of such a useful method would be?
```

## `Object.{pick, omit}`
> ECMAScript Proposal, specs, and reference implementation for `Object.pick`, `Object.omit`.

Spec drafted by [@Aleen](https://github.com/aleen42).

### Motivation

> Let us consider few scenarios from the real would to understand what we are trying to solve in this propsoal. 

Reviewer 2: ```suggestion
Let us consider a few scenarios from the real world to understand what we are trying to solve in this proposal. 
```

* Checking if a component `shouldReload` by extracting `compareKeys` from `props` and compare it with `prevProps`. 
* Say we have a `depsObject` and we need to ignore all `@internal/packages` from it.
* We have `props` from which we need to remove `[‘_csrf’, ‘_method’]`
* We need to construct a `newModelData` by removing `action.deleted` from `({ ...state.models, ...action.update })`
* Filtering configuration objects when the filter list is given by a `CLI` argument.

Well, you see life is all about `pick`ing what we want and `omit`ing what we don't! 

Would life be easier if the language provided a convenient method to help us during similar scenarios?

Now, one might argue saying we can implement `pick` and `omit` as below:

```js
const pick = (obj, keys) => Object.fromEntries(
    keys.map(k => obj.hasOwnProperty(k) && [k, obj[k]]).filter(x => x)
);

/*
We can also use a Destructuring assignment
const { authKey, ...toLog } = userInfo;
*/
```

```js
const omit = (obj, keys) => Object.fromEntries(
    keys.map(k => !obj.hasOwnProperty(k) && [k, obj[k]]).filter(x => x)
);
```

The major challenges we see with the above implementations:

* It is not ergonomic!

> * If we opt the destructuring way it doesn't work at all for `pick`, or for `omit` with dynamic values.

Reviewer 2: ```suggestion
* If we opt for the destructuring way it doesn't work at all for `pick`, or for `omit` with dynamic values.
```

We can read more about such use-cases and challenges from `es.discourse` below:

* [Object.{pick,omit}](https://es.discourse.group/t/object-prototype-pick-object-prototype-omit/515).
* [Object restructuring syntax](https://es.discourse.group/t/object-restructuring-syntax/651)
* [Object Array Pick](https://es.discourse.group/t/object-array-pick/992)
* [js-pick-notation](https://github.com/rtm/js-pick-notation)
* [slect multiple object values](https://es.discourse.group/t/set-multiple-object-values-a-b-undefined/1052/3)

With that in mind would it not be easier if we had `Object.pick` and `Object.omit` static methods?!

> Let us now discuss about what the API of such a useful method would be?

Reviewer 2: ```suggestion
Let us now discuss what the API of such a helpful method would be?
```

The major challenges we see with the above implementations:

* It is not ergonomic 

> * We can't omit for unknown values (with destructuring)

Reviewer: the thing to convey tho is that this only works for "omit" with static values. destructuring doesn't work at all for "pick", or for "omit" with dynamic values.

Author: nods, rephrased the text.

Reviewer 2: Like something:

1. Destructuring cannot clone a new object while `Object.pick` can
2. Destructuring cannot pick up properties from the prototype while `Object.pick` can
3. Destructuring cannot pick properties dynamically, while `Object.pick` can
4. Destructuring cannot omit some properties, and we can only clone and delete without this proposal
5. ...
